### PR TITLE
Run a proxy for ironic when provisioning network is disabled

### PR DESCRIPTION
**Pull request by dtantsur**
Without a provisioning network, ICC uses IP address of the control plane
node with the metal3 pod. If the pod relocates, the ramdisk will lose
access to Ironic. To solve this issue, run a simple HTTP(s) reverse proxy
as a DaemonSet on all control plane nodes.

### CODE REVIEW

File: provisioning/utils.go [hunk 1]
```diff
@@ -47,8 +48,27 @@ func getPodHostIP(podClient coreclientv1.PodsGetter, targetNamespace string) (st
...
-func GetIronicIP(client kubernetes.Interface, targetNamespace string, config *metal3iov1alpha1.ProvisioningSpec) (string, error) {
-	if config.ProvisioningNetwork != metal3iov1alpha1.ProvisioningNetworkDisabled && !config.VirtualMediaViaExternalNetwork {
+func GetServerInternalIP(osclient osclientset.Interface) (string, error) {
+	infra, err := osclient.ConfigV1().Infrastructures().Get(context.Background(), "cluster", metav1.GetOptions{})
+	if err != nil {
+		err = fmt.Errorf("Cannot get the 'cluster' object from infrastructure API: %w", err)
+		return "", err
+	}
+
+	// NOTE(dtantsur): do we need to handle non-BareMetal platforms here?
```
Comment (dtantsur) [inline provisioning/utils.go:58]: I guess we do. OpenStack and vSphere have the same field, for None we can fall back to the previous process.

I suggest we do it in a follow-up to reduce the scope of this (already large) change.

File: provisioning/utils.go [hunk 2]
```diff
@@ -47,8 +48,27 @@ func getPodHostIP(podClient coreclientv1.PodsGetter, targetNamespace string) (st
...
+	if infra.Status.PlatformStatus.BareMetal == nil {
+		err = fmt.Errorf("Cannot detect server API VIP: not a baremetal platform")
+		return "", err
+	}
+
+	// FIXME(dtantsur): handle the new APIServerInternalIPs field and the dualstack case.
```
Comment (dtantsur) [inline provisioning/utils.go:64]: Can be a follow-up as well IMO to avoid updating vendored dependencies in the same change.

**Comment by dtantsur** [comment]
/hold

I see a problem here (not sure why the CI fails to catch it): ICC reuses the same IP for inspector.

**Comment by dtantsur** [comment]
Now requires https://github.com/openshift/image-customization-controller/pull/60

**Comment by rhjanders** [comment]
> /hold
> 
> I see a problem here (not sure why the CI fails to catch it): ICC reuses the same IP for inspector.

I wonder if maybe it worked by a co-incidence? (IPs worked out to be the same, or inspector was listening on * and it worked in this case, or something) It's been puzzling me too why CI didn't pick this up.

**Comment by dtantsur** [comment]
Okay, this works in my testing. Once the ICC change merges, we can check the virtualmedia job and unhold the PR.

**Comment by dtantsur** [comment]
I suspect we need to rerun the job once the ICC change is in containers.

File: provisioning/ironic_proxy.go [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+			UpdateStrategy: appsv1.DaemonSetUpdateStrategy{
+				RollingUpdate: &appsv1.RollingUpdateDaemonSet{
+					MaxUnavailable: &maxUnavail,
+				},
+			},
+		},
+	}, nil
+}
+
+func UseIronicProxy(config *metal3iov1alpha1.ProvisioningSpec) bool {
```
Comment (honza) [inline provisioning/ironic_proxy.go:198]: nit: `s/UseIronicProxy/ShouldUseIronicProxy/`

File: provisioning/utils.go (hunk 2 above)
Comment (honza) [reply provisioning/utils.go:64]: :+1:

**Pull request merged**